spotbugs-annotations/build.gradle

```diff
@@ -7,7 +7,7 @@ apply from: "$rootDir/gradle/javadoc.gradle"
 apply from: "$rootDir/gradle/maven.gradle"
 
 dependencies {
-  compile 'com.google.code.findbugs:jsr305:3.0.1'
+  compile 'com.google.code.findbugs:jsr305:3.0.2'
```

Comment:
refer https://mailman.cs.umd.edu/pipermail/findbugs-discuss/2017-March/004354.html

Reply:
Any jar that has classes in `java` or `javax` packages violates the Oracle JVM redistribution license. By adding back such a dependency you restore this limitation

Reply:
OK then delete `ReturnValuesAreNonnullByDefault` and ask users to migirate to `DefaultAnnotation` or others :+1:

Reply:
Not only `ReturnValuesAreNonnullByDefault` but also `CheckForNull` needs migration, because current spotbugs-annotation does not support `When`. and `TypeQualifierNickname`.

If the distribution is the problem, we can change this dependency to `compileOnly`, then our distribution package will not include `jsr-305-3.0.1.jar`.

Reply:
Oops, `TypeQualifierAnnotation` depends on `When` in jsr-305...

Reply:
Note: here is license http://www.oracle.com/technetwork/java/javase/terms/license/index.html

> F. JAVA TECHNOLOGY RESTRICTIONS. You may not create, modify, or change the behavior of, or authorize your licensees to create, modify, or change the behavior of, classes, interfaces, or subpackages that are in any way identified as "java", "javax", "sun", “oracle” or similar convention as specified by Oracle in any naming convention designation.

IANAL but who violate this is not us but jsr305.jar, is it also bad to depend on it? What do you mean by "you restore this limitation"?

Reply:
When a class is loaded, if the annotation is not present... no foul...

Now what happens if there is a reboot JSR that superseded JSR 305, and it adds `javax.annotation.Nonnull`, but decided to use `value` for the when clause so that it is `@Nonnull(___)` rather than forcing everyone to type in `@Nonnull(when=___)`... save 5 characters many times... that's a win...

In that case when a class is loaded... and that includes an annotation... you will have the *entire* class fail to load as the annotation is present but the signature is invalid.

You have to be exceedingly careful in how you evolve the fields in an annotation.

Oracle are aware of this issue, which is why they have that restriction, because if they do not keep the guard of the "must be published as part of a JSR process" then it would be *their* fault for releasing a JSR-xyz that had the alternative `javax.annotation.Nonnull`.

Right now, however, there is no publication from JSR-305, so it will be your fault if you have those class references in your class files.

Additionally, a really nit-picking lawyer told me that the classloading failure described above could be interpreted as "change the behavior of"... so if you are distributing an oracle JVM, then - in his overly pedantic view - you should not even have those annotations referenced in your class files... 

now that last bit is not an argument that I personally put much strength in... but I do think that at some point in time there may be a JSR-305 reboot, and we do not know what they would produce from that, hence we should not be preparing to shoot ourselves in the foot, especially given that the IDE that redistribute a JVM in their installers have gone to great lengths not to include those annotations in their classfiles... indicating to me that they share somewhat similar concerns

Reply:
In my personal opinion, we can keep current dependency for 3.1.0 release, to keep SpotBugs as much as same with FindBugs. Of course, we all know this dependency is negative legacy, we should have issue for 4.0.0 to delete them.